For the below data write the forward pass through ANN using python and numpy.
Pass the data through the network, display the outputs and error
Data is about frequency(cell) of words(columns) present in a mail(row). Need to classify the mail as spam or not.
	Dear 	Dinner	Money	Spam – 1 Not spam - 0

Mail 1	1	2	1	0

Mail 2	1	0	4	1

Mail 3 	0	3	0	0

Mail4	1	0	2	1


Note: Use proper activation functions.
          Assume weights and biases.
          Upload only the .ipynb file without zipping.



In [18]:
import numpy as np
import tensorflow as tf
from tensorflow import keras

Given architecture of the neural network

In [30]:
input_size = 3
output_size = 1
hidden_size = 4

### Sigmoid activation function

In [ ]:
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

### inputs and expected outputs

In [ ]:
inputs = np.array([[1, 2, 1,0], [1, 0, 4,1], [0, 3, 0,1], [1, 0, 2,1]])

expected_outputs = np.array([[0], [1], [0], [1]])

spam-1

not spam-0

### weights and bias for the ouput layers

In [36]:
output_weights = np.random.uniform(size=(4, 1))
output_bias = np.random.uniform(size=(1, 1))


### weights and bias for the hidden layers

In [37]:
hidden_weights = np.random.uniform(size=(4, 4))
hidden_bias = np.random.uniform(size=(1, 4))

### Train the neural network

In [ ]:
# Perform the forward pass through the network

In [40]:
ab = sigmoid(np.dot(inputs, weights_input_hidden) + bias_hidden)
bc = sigmoid(np.dot(hidden_layer, weights_hidden_output) + bias_output)

In [41]:
print(bc)

[[0.63713105]
 [0.57598141]
 [0.65480363]
 [0.5800638 ]]


## INFERENCE:
    We conclude that the output values Mail3 has highest spam.mail3(0.6548)

### Calculate the error

In [50]:
output_error = expected_output - output_layer
hidden_error = np.dot(output_error, weights_hidden_output.T) * hidden_layer * (1 - hidden_layer)

In [52]:
 print(output_error)

[[-0.63713105]
 [ 0.42401859]
 [-0.65480363]
 [ 0.4199362 ]]


## INFERENCE:

Mail2(0.42401859) and  Mail4(0.4199362) are positive values so that  spam. Mail1(-0.63713105) and Mail3(-0.65480363) are negative values so that is not spam.